### PR TITLE
DM-14227 Fix FITS table load failure and multiple images display in one window issue

### PR DESCRIPTION
**Pull request by cwang2016**
This development fixes, 
- the  fits table load failure issue by updating some global setting variable related to 'long string enabled' for fits reading before using starlink to load fits table since the value related to 'long string enabled' is possibly changed the fits image query issued after previous fits table loading.
- the error while rendering multiple images in one window. 

test:
pull firefly_test_data
start firefly, do uploading, and try the fits test files under firefly_test_data/FileUpload-samples/fits. the problematic sample related to this fixing is firefly_test_data/FileUpload-samples/fits/problemFits/src.fits.

### CODE REVIEW

**Comment by cwang2016** [comment]
Update the code to make the use of fits library (nom.tam.fits) and starlink thread safe while reading the fits header and load fits table.

File: src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java [hunk 1]
```diff
@@ -615,10 +604,9 @@ private static String getAttributeValue(Object obj,
      * @param catName data group title (catName is not used).
      * @deprecated
      */
-    public static List<DataGroup> convertFITSToDataGroup(String FITS_filename, String catName)
+    public static List<DataGroup> convertFITSToDataGroup(String FITS_filename, String catName, int idx)
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java]: Similar to `getStarTableList`, this should be done via an overloaded method.  i.e.
`public static DataGroup convertFITSToDataGroup(String FITS_filename, String catName, int idx)` 

If the original method is not used and cannot be easily implemented, I would just throw an unsupported exception for now.

File: src/firefly/java/edu/caltech/ipac/firefly/server/util/ipactable/DataGroupReader.java [hunk 2]
```diff
@@ -46,9 +46,9 @@ public static DataGroup readAnyFormat(File inf, int tableIndex) throws IOExcepti
         } else if (format == Format.FITS ) {
             try {
                 // Switch to the new function:
-                List<DataGroup> retval = FITSTableReader.convertFitsToDataGroup(inf.getAbsolutePath(), null, null, FITSTableReader.DEFAULT);
+                List<DataGroup> retval = FITSTableReader.convertFitsToDataGroup(inf.getAbsolutePath(), null, null, FITSTableReader.DEFAULT, tableIndex);
                 if (retval != null && retval.size() > 0) {
-                    return retval.get(tableIndex);
+                    return retval.get(0);
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/firefly/server/util/ipactable/DataGroupReader.java]: This section will use the new overloaded method.  It'll be cleaner.

File: src/firefly/java/edu/caltech/ipac/util/FitsHDUUtil.java [hunk 3]
```diff
@@ -115,13 +115,16 @@ public static DataGroup fitsHeaderToDataGroup(String fitsFile) {
         DataGroup dg = new DataGroup("fits", cols);
         String invalidMsg = "invalid fits file";
 
+        FitsFactory.useThreadLocalSettings(true);
+        boolean prevLongStringEnabled = FitsFactory.isLongStringsEnabled();
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/util/FitsHDUUtil.java]: The use of `useTreadLocalSettings` is not needed here.  The added code can be removed.

File: src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java [hunk 4]
```diff
@@ -210,53 +209,43 @@ else if (strategy == FULLY_FLATTEN) {
...
      * Get a list of star tables from a FITS file.
      * @param fits_filename
+     * @param table_idx
      * @return List of star tables.
      * @throws IOException
      */
-    public static List getStarTableList(String fits_filename) {
 
-        ArrayList<StarTable> tList = new ArrayList<>();
+    public static List getStarTableList(String fits_filename, int table_idx) {
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java]: I understand the need for this change. But, it should be a new method.  i.e.
`public static StarTable getStarTable(String fits_filename, int table_idx) `

If the original method is not used and cannot be easily implemented, I would just throw an unsupported exception for now.

File: src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java [hunk 5]
```diff
@@ -210,53 +209,43 @@ else if (strategy == FULLY_FLATTEN) {
...
+
+        ArrayList<StarTable> tList =  new ArrayList<>();
 
-        // disable long string for HeaderCard create while collecting tables from StarTableFactory to work around the exception error
-        // sent from nom.tam.fits.
+        // disable long string for HeaderCard creation while collecting table with table_idx from StarTableFactory to work around
+        // the exception error sent from nom.tam.fits.
 
         FitsFactory.useThreadLocalSettings(true);
+        boolean prevLongStringEnabled = FitsFactory.isLongStringsEnabled();
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java]: No need to save/recover that one variable.  You can set `FitsFactory.useThreadLocalSettings(false);` at the end of this method to clear all threadlocal settings.

**Comment by cwang2016** [comment]
the use of thread local setting and long string enabling are updated per review. 
the unused jars for starlink (stil. jar, regclient.jar) are removed.

File: src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java [hunk 6]
```diff
@@ -193,70 +192,87 @@ else if (strategy == FULLY_FLATTEN) {
...
+
+    /**
+     * Get a star tables from a FITS file.
+     * @param fits_filename
+     * @param table_idx
+     * @return one star table or none
+     * @throws IOException
+     */
+
+    public static StarTable getStarTableList(String fits_filename, int table_idx) {
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/astro/FITSTableReader.java]: Really minor.  Will you rename the method to getStarTable?  It is not returning a list anymore.
Feel free to merge after your change.

**Pull request merged**